### PR TITLE
xcube serve does not provide datasets after service configuration changes

### PR DESCRIPTION
**Pull request by TonioF**
Solves #678. The changes in this PR ensure that there is always a DataStorePool in the service context, and that the dataset_configs correspond to its content. The DataStorePool is repopulated when it has been emptied, which usually happens after an external config change.

Checklist:

* [ ] Add unit tests and/or doctests in docstrings
* ~[ ] Add docstrings and API docs for any new/modified user-facing classes and functions~
* ~[ ] New/modified features documented in `docs/source/*`~
* [x] Changes documented in `CHANGES.md`
* [ ] AppVeyor CI passes
* [ ] Test coverage remains or increases (target 100%)

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -22,6 +22,10 @@
 
 * Fixed ESA CCI example notebook. (#680)
 
+* `xcube serve` now provides datasets after changes of the service 
+  configuration.
```
Comment (forman) [inline CHANGES.md:26]:
```suggestion
  configuration while the server is running.
```

File: CHANGES.md [hunk 2]
```diff
@@ -22,6 +22,10 @@
...
+  Previously, it was necessary to restart a service to load the changes. (#678)
```
Comment (forman) [inline CHANGES.md:27]:
```suggestion
  Previously, it was necessary to restart the server to load the changes. (#678)
```

File: xcube/webapi/context.py [hunk 3]
```diff
@@ -303,16 +302,12 @@ def get_variable_for_z(self, ds_id: str, var_name: str, z_index: int) -> xr.Data
...
-        if self._dataset_configs is None:
-            with self._lock:
-                dataset_configs = self._config.get('Datasets', [])
-                dataset_configs += \
-                    self.get_dataset_configs_from_stores()
-                self._dataset_configs = dataset_configs
-                self._maybe_assign_store_instance_ids()
+        with self._lock:
+            if self._dataset_configs is None:
+                self._set_up_data_store_pool_and_dataset_configs()
```
Comment (forman) [inline xcube/webapi/context.py:307]: Only lock, if needed. Double-check to be on the safe side (see https://en.wikipedia.org/wiki/Double-checked_locking).

```suggestion
        if self._dataset_configs is None:
            with self._lock:
                if self._dataset_configs is None:
                    self._set_up_data_store_pool_and_dataset_configs()
```

File: xcube/webapi/context.py [hunk 4]
```diff
@@ -499,24 +486,34 @@ def new_dataset_metadata(self,
         # noinspection PyTypeChecker
         return dataset_metadata
 
-    def get_data_store_pool(self) -> Optional[DataStorePool]:
+    def get_data_store_pool(self) -> DataStorePool:
+        with self._lock:
```
Comment (forman) [inline xcube/webapi/context.py:490]: See above, use double-checked locking.

**Review by forman — CHANGES_REQUESTED** [review]
Good! But please update according to my suggestions.

**Review by AliceBalfanz — APPROVED** [review]
From a pure usability aspect, this now works well. Tested locally.

**Pull request merged**